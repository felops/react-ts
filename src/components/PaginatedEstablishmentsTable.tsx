import { useState } from "react";
import { EstablishmentsTable } from "./EstablishmentsTable";
import { EstablishmentsTableNavigation } from "./EstablishmentsTableNavigation";
import { getEstablishmentRatings, getEstablishmentRatingsByAuthority } from "../api/ratingsAPI";
import { useQuery } from "@tanstack/react-query";
import { getAuthorities } from "../api/authoritiesAPI";

export const PaginatedEstablishmentsTable = () => {
  const [pageNum, setPageNum] = useState(1);
  const [pageCount] = useState(100);
  const [localAuthorityId, setLocalAuthorityId] = useState('');
  const { data: authoritiesData } = useQuery({
    queryKey: ['authorities'],
    queryFn: () => getAuthorities()
  });
  const { isFetching, data, error } = useQuery({
    queryKey: ['establishments', localAuthorityId, pageNum],
    queryFn: () => localAuthorityId
      ? getEstablishmentRatingsByAuthority(localAuthorityId, pageNum)
      : getEstablishmentRatings(pageNum)
  });

  const onChangeLocalAuthority = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setLocalAuthorityId(event.target.value);
    setPageNum(1);
  };

  const handlePreviousPage = async () => {
    pageNum > 1 && setPageNum(pageNum - 1);
  }

  const handleNextPage = async () => {
    pageNum < pageCount && setPageNum(pageNum + 1);
  }

  if (error) {
    return <div>Error: please, try again later</div>;
  }

  return (
    <>
      <select onChange={onChangeLocalAuthority} value={localAuthorityId}>
        <option value="">All Authorities</option>
        {authoritiesData?.authorities?.map((authority) => (
          <option key={authority.LocalAuthorityId} value={authority.LocalAuthorityId}>
            {authority.Name}
          </option>
        ))}
      </select>
      <EstablishmentsTable
        establishments={data?.establishments}
        isLoading={isFetching}
      />
      <EstablishmentsTableNavigation
        pageNum={pageNum}
        pageCount={pageCount}
        onPreviousPage={handlePreviousPage}
        onNextPage={handleNextPage}
      />
    </>
  );
};
