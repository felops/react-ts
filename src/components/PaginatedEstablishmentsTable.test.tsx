import { render, screen } from '@testing-library/react';
import { useQuery } from '@tanstack/react-query';
import { PaginatedEstablishmentsTable } from './PaginatedEstablishmentsTable';

jest.mock('@tanstack/react-query', () => ({
  ...jest.requireActual("@tanstack/react-query"),
  useQuery: jest.fn(),
}));

const mockAuthorityData = [
  {
    LocalAuthorityId: "1",
    Name: "Authority 1"
  },
  {
    LocalAuthorityId: "2",
    Name: "Authority 2"
  },
  {
    LocalAuthorityId: "3",
    Name: "Authority 3"
  },
];

const mockEstablishmentsData = [
  {
    BusinessName: "Joe",
    RatingValue: "5"
  },
  {
    BusinessName: "Habib",
    RatingValue: "4"
  },
  {
    BusinessName: "Mc Hogwarts",
    RatingValue: "3"
  },
]

describe("PaginatedEstablishmentsTable", () => {
  it('renders correctly when is loading', () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: { authorities: mockAuthorityData },
      error: null,
    });
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: true,
      data: { establishments: mockEstablishmentsData },
      error: null,
    });
    const { asFragment } = render(<PaginatedEstablishmentsTable />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when is NOT loading', () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: { authorities: mockAuthorityData },
      error: null,
    });
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: { establishments: mockEstablishmentsData },
      error: null,
    });
    const { asFragment } = render(<PaginatedEstablishmentsTable />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display loading state', async () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: { authorities: mockAuthorityData },
      error: null,
    });
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: true,
      data: { establishments: mockEstablishmentsData },
      error: null,
    });
    render(<PaginatedEstablishmentsTable />);

    const loading = await screen.getByText('Loading...');
    expect(loading).toBeTruthy();
  });

  it('should NOT display loading state', async () => {
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: { authorities: mockAuthorityData },
      error: null,
    });
    (useQuery as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      data: { establishments: mockEstablishmentsData },
      error: null,
    });
    render(<PaginatedEstablishmentsTable />);

    const loading = await screen.queryAllByText('Loading...');
    expect(loading).toEqual([]);
  });
});
