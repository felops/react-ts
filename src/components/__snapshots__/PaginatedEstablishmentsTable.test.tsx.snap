// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedEstablishmentsTable renders correctly when is NOT loading 1`] = `
<DocumentFragment>
  <label>
    Filter by Authority:
  </label>
  <select>
    <option
      value=""
    >
      All Authorities
    </option>
    <option
      value="1"
    >
      Authority 1
    </option>
    <option
      value="2"
    >
      Authority 2
    </option>
    <option
      value="3"
    >
      Authority 3
    </option>
  </select>
  <table
    style="min-height: 25rem;"
  >
    <thead>
      <tr>
        <th />
        <th>
          Business Name
        </th>
        <th>
          Rating Value
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <input
            type="checkbox"
          />
        </td>
        <td
          class="font-size-20"
        >
          <a
            data-discover="true"
            href="/22999/details"
          >
            Joe
          </a>
        </td>
        <td
          class="font-size-20"
        >
          5
        </td>
      </tr>
      <tr>
        <td>
          <input
            type="checkbox"
          />
        </td>
        <td
          class="font-size-20"
        >
          <a
            data-discover="true"
            href="/67890/details"
          >
            Habib
          </a>
        </td>
        <td
          class="font-size-20"
        >
          4
        </td>
      </tr>
      <tr>
        <td>
          <input
            type="checkbox"
          />
        </td>
        <td
          class="font-size-20"
        >
          <a
            data-discover="true"
            href="/09888/details"
          >
            Mc Hogwarts
          </a>
        </td>
        <td
          class="font-size-20"
        >
          3
        </td>
      </tr>
    </tbody>
  </table>
  <nav>
    <button
      disabled=""
      type="button"
    >
      &lt;  previous
    </button>
    <span
      style="font-weight: bold; margin: 1rem;"
    >
      1
    </span>
    <button
      type="button"
    >
      next  &gt;
    </button>
  </nav>
</DocumentFragment>
`;

exports[`PaginatedEstablishmentsTable renders correctly when is loading 1`] = `
<DocumentFragment>
  <label>
    Filter by Authority:
  </label>
  <select>
    <option
      value=""
    >
      All Authorities
    </option>
    <option
      value="1"
    >
      Authority 1
    </option>
    <option
      value="2"
    >
      Authority 2
    </option>
    <option
      value="3"
    >
      Authority 3
    </option>
  </select>
  <table
    style="min-height: 25rem;"
  >
    <thead>
      <tr>
        <th />
        <th>
          Business Name
        </th>
        <th>
          Rating Value
        </th>
      </tr>
      <tr>
        <th
          colspan="3"
        >
          <div>
            Loading...
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <input
            type="checkbox"
          />
        </td>
        <td
          class="font-size-20"
        >
          <a
            data-discover="true"
            href="/22999/details"
          >
            Joe
          </a>
        </td>
        <td
          class="font-size-20"
        >
          5
        </td>
      </tr>
      <tr>
        <td>
          <input
            type="checkbox"
          />
        </td>
        <td
          class="font-size-20"
        >
          <a
            data-discover="true"
            href="/67890/details"
          >
            Habib
          </a>
        </td>
        <td
          class="font-size-20"
        >
          4
        </td>
      </tr>
      <tr>
        <td>
          <input
            type="checkbox"
          />
        </td>
        <td
          class="font-size-20"
        >
          <a
            data-discover="true"
            href="/09888/details"
          >
            Mc Hogwarts
          </a>
        </td>
        <td
          class="font-size-20"
        >
          3
        </td>
      </tr>
    </tbody>
  </table>
  <nav>
    <button
      disabled=""
      type="button"
    >
      &lt;  previous
    </button>
    <span
      style="font-weight: bold; margin: 1rem;"
    >
      1
    </span>
    <button
      type="button"
    >
      next  &gt;
    </button>
  </nav>
</DocumentFragment>
`;
